otherlibs/cram/bin/run.ml

```diff
@@ -46,133 +56,264 @@ let run_expect_test file ~f =
+         into a file and then source it in the main script. *)
+      let user_shell_code_file = file ~ext:".sh" in
+      let user_shell_code_file_sh_path = sh_path user_shell_code_file in
+      cat_eof oc lines ~dest:user_shell_code_file;
+      (* Where we store the output of shell code written by the user *)
+      let user_shell_code_output_file = file ~ext:".output" in
+      let user_shell_code_output_file_sh_path =
+        sh_path user_shell_code_output_file
       in
-      translate_path_for_sh prog |> quote_for_sh
+      fprln oc ". %s > %s 2>&1" user_shell_code_file_sh_path
+        user_shell_code_output_file_sh_path;
+      fprln oc {|printf "$?\0$%s\0" >> %s|} _BUILD_PATH_PREFIX_MAP
+        metadata_file_sh_path;
+      Command { command = lines; output_file = user_shell_code_output_file }
+  in
+  let cram_to_output = List.mapi ~f:loop cram_stanzas in
+  close_out oc;
+  { script; cram_to_output; metadata_file }
+
+let display_with_bars s =
+  List.iter (String.split_lines s) ~f:(Printf.eprintf "| %s\n")
+
+let run ~sanitizer ~file lexbuf =
+  let temp_dir =
+    let suffix = Filename.basename file in
+    Temp.dir ~prefix:".dune.cram" ~suffix
   in
+  let cram_stanzas = cram_stanzas lexbuf in
+  let sh_script = create_sh_script cram_stanzas ~temp_dir in
   Sys.chdir (Filename.dirname file);
   let cwd = Sys.getcwd () in
-  Unix.putenv "LC_ALL" "C";
-  extend_build_path_prefix_map ~cwd;
-  let script, oc = open_temp_file ".sh" in
-  let prln fmt = Printf.fprintf oc (fmt ^^ "\n") in
-  (* Shell code written by the user might not be properly terminated. For
-     instance the user might forgot to write [EOF] after a [cat <<EOF]. If we
-     wrote this shell code directly in the main script, it would hide the rest
-     of the script. So instead, we dump each user written shell phrase into a
-     file and then source it in the main script. *)
-  let user_shell_code_file = temp_file ".sh" in
-  let user_shell_code_file_sh_path =
-    translate_path_for_sh user_shell_code_file
-  in
-  (* Where we store the output of shell code written by the user *)
-  let user_shell_code_output_file = temp_file ".output" in
-  let user_shell_code_output_file_sh_path =
-    translate_path_for_sh user_shell_code_output_file
-  in
-
-  (* Produce the following shell code:
-
-     {v cat <<"EOF_<n>" <data> EOF_<n> v}
-
-     choosing [<n>] such that [CRAM_EOF_<n>] doesn't appear in [<data>]. *)
-  let cat_eof ?dest lines =
-    let rec loop n =
-      let sentinel =
-        if n = 0 then
-          "EOF"
-        else
-          sprintf "EOF%d" n
-      in
-      if List.mem sentinel ~set:lines then
-        loop (n + 1)
-      else
-        sentinel
+  let env =
+    let env = Env.initial in
+    let env = Env.add env ~var:"LC_ALL" ~value:"C" in
+    let env = extend_build_path_prefix_map ~env ~cwd in
+    let env =
+      Env.add env ~var:"TMPDIR" ~value:(Path.to_absolute_filename temp_dir)
```

Comment:
It's TEMP on Windows

Reply:
BTW, since we are also creating files in this directory, we should give the command a sub-directory of `temp_dir` to avoid clashes between our files and the command temporary files.

Reply:
Done.